### PR TITLE
upgrade ginkgo to ginkgo/v2

### PR DESCRIPTION
**Pull request by ruquanzhao**
Signed-off-by: Ruquan Zhao <ruquan.zhao@arm.com>

Since Ginkgo has been migrated from v1 to v2, [ginkgo_MIGRATING_TO_V2](https://onsi.github.io/ginkgo/MIGRATING_TO_V2),
How about we upgrade it in this project?

And this blocks kubernetes[#109111](https://github.com/kubernetes/kubernetes/pull/109111), which is upgrading ginkgo to v2 in kubernetes.

### CODE REVIEW

**Comment by chendave** [comment]
@liggitt , I think this pr is enough, per my investigation on the custom report for printing the new line, there is no need to migrate that report.

@ruquanzhao could you pls open this pr for review?

**Comment by chendave** [comment]
something weird is that there is no jobs need to be run for this pr? or this test jobs will visible after removing the W-I-P label?

**Comment by liggitt** [comment]
/assign @apelisse @lavalamp

I'm not sure what the story is around CI for this repo... I'm also surprised there's no CI jobs running

**Comment by lavalamp** [comment]
I don't know anything about the test setup in this repo. @apelisse or @seans3 maybe?

I don't have any particular objection to upgrading this.

**Comment by liggitt** [comment]
> Yeah, looks good. The CI wasn't running because OP wasn't approved to run the tests, they should be running now.

huh... does that mean there's a gap where PRs can merge without CI running?

**Comment by apelisse** [comment]
Well, it's too late now, but we should try to see if tide would consider the PR to be merged without the approval.

**Comment by chendave** [comment]
@ruquanzhao  this should be fixed.

```
BeforeSuite(func() {
/home/runner/work/kube-openapi/kube-openapi/src/k8s.io/kube-openapi/test/integration/integration_suite_test.go:73
  It looks like you are trying to add a [BeforeSuite] node within a container
  node.
BeforeSuite can only be called at the top level.
```

I don't think we can merge this before https://github.com/kubernetes/kubernetes/pull/109111, am I right? @liggitt ?

**Comment by liggitt** [comment]
> I don't think we can merge this before https://github.com/kubernetes/kubernetes/pull/109111, am I right? @liggitt ?

this repo is independent… stuff merges here, then the version of this repo used by k/k gets bumped in a PR there

**Comment by ruquanzhao** [comment]
I pushed some codes to pass the CI fails before.

But the CI doesn't start. Can you help with it? @apelisse

**Comment by apelisse** [comment]
Trjggered!

**Comment by apelisse** [comment]
Code looks good otherwise, just curious about these new dependencies.

File: go.sum [hunk 1]
```diff
@@ -9,6 +9,9 @@ github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578/go.mod h1:uGdko
 github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a h1:idn718Q4B6AGu/h5Sxe66HYVdqdGu2l9Iebqhi/AEoA=
 github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a/go.mod h1:lB+ZfQJz7igIIfQNfa7Ml4HSf2uFQQRzpGGRXenZAgY=
 github.com/census-instrumentation/opencensus-proto v0.2.1/go.mod h1:f6KPmirojxKA12rnyqOA5BBL4O983OfeGPqjHWSTneU=
+github.com/chzyer/logex v1.1.10/go.mod h1:+Ywpsq7O8HXn0nuIou7OrIPyXbp3wmkHB+jjWRnGsAI=
+github.com/chzyer/readline v0.0.0-20180603132655-2972be24d48e/go.mod h1:nSuG5e5PlCu98SY8svDHJxuZscDgtXS6KTTbou5AhLI=
+github.com/chzyer/test v0.0.0-20180213035817-a1ea475d72b1/go.mod h1:Q3SI9o4m/ZMnBNeIyt5eFwwo7qiLfzFZmjNmxjkiQlU=
```
[the remark this answers is not in the document]
Comment (liggitt) [reply go.sum:14]: I'm not thrilled with the additions, and I'm a little surprised ginkgo v1 is still showing up here

[the remark this answers is not in the document]
Comment (chendave) [reply go.sum:14]: Looks like a legacy leftovers, just like `github.com/golang/protobuf`, @ruquanzhao maybe we can remove Ginkgo v1 manually from this file and try to test it again? I am not sure is there any tool (script) helps us to do that.

**Comment by apelisse** [comment]
ginkgo and gomega are just having fun I think, but the ginkgo v1 dependency is here to stay I would think:
```
> go mod graph | grep ginkgo | grep gomega
github.com/onsi/ginkgo/v2@v2.1.4 github.com/onsi/gomega@v1.19.0
github.com/onsi/gomega@v1.19.0 github.com/onsi/ginkgo/v2@v2.1.3
github.com/onsi/ginkgo/v2@v2.1.3 github.com/onsi/gomega@v1.17.0
github.com/onsi/gomega@v1.17.0 github.com/onsi/ginkgo@v1.16.4
github.com/onsi/ginkgo@v1.16.4 github.com/onsi/gomega@v1.10.1
github.com/onsi/gomega@v1.10.1 github.com/onsi/ginkgo@v1.12.1
github.com/onsi/ginkgo@v1.12.1 github.com/onsi/gomega@v1.7.1
github.com/onsi/gomega@v1.7.1 github.com/onsi/ginkgo@v1.6.0
```

**Comment by liggitt** [comment]
/hold

We originally thought the k/k PR switching to ginkgo/v2 was only pulling in v1 because of the kube-openapi dep on it (which is what prompted this PR). If there's a more tangled story around ginkgo/v2 → gomega → gingko/v1, that should be resolved before bumping to ginkgo/v2 in any of our deps

**Comment by apelisse** [comment]
I'm trying to see where those dependencies are coming from:
```
# ginkgo/v2 needs google/pprof
github.com/onsi/ginkgo/v2@v2.1.4 github.com/google/pprof@v0.0.0-20210407192527-94a9f03dee38
# google/pprof adds these 4 dependencies:
github.com/google/pprof@v0.0.0-20210407192527-94a9f03dee38 github.com/chzyer/logex@v1.1.10
github.com/google/pprof@v0.0.0-20210407192527-94a9f03dee38 github.com/chzyer/readline@v0.0.0-20180603132655-2972be24d48e
github.com/google/pprof@v0.0.0-20210407192527-94a9f03dee38 github.com/chzyer/test@v0.0.0-20180213035817-a1ea475d72b1
github.com/google/pprof@v0.0.0-20210407192527-94a9f03dee38 github.com/ianlancetaylor/demangle@v0.0.0-20200824232613-28f6c0f3b639
```

gomega 1.7 is also now explicitely pulled and pulls all the other added dependencies basically.

**Comment by liggitt** [comment]
I think bumping the `go.mod` to indicate `go 1.18` and redoing `go mod tidy` is needed to pick up the transitive pruning improvements in go 1.18

**Comment by liggitt** [comment]
that would eliminate a lot of the noise from go.sum

**Comment by apelisse** [comment]
You can rebase this on top of master and you should be good to go.

File: pkg/util/proto/openapi_suite_test.go [hunk 2]
```diff
@@ -35,15 +35,15 @@ func TestOpenapi(t *testing.T) {
...
-func (newlineReporter) SpecWillRun(specSummary *SpecSummary) {}
+func (newlineReporter) SpecWillRun(specSummary *types.SpecSummary) {}
 
-func (newlineReporter) SpecDidComplete(specSummary *SpecSummary) {}
+func (newlineReporter) SpecDidComplete(specSummary *types.SpecSummary) {}
 
-// SpecSuiteDidEnd Prints a newline between "35 Passed | 0 Failed | 0 Pending | 0 Skipped" and "--- PASS:"
-func (newlineReporter) SpecSuiteDidEnd(summary *SuiteSummary) { fmt.Printf("\n") }
+// SuiteDidEnd Prints a newline between "35 Passed | 0 Failed | 0 Pending | 0 Skipped" and "--- PASS:"
+func (newlineReporter) SuiteDidEnd(summary *types.SuiteSummary) { fmt.Printf("\n") }
```
Comment (chendave) [inline pkg/util/proto/openapi_suite_test.go:49]: you can just remove this newline reporter If you want,  pls see: https://github.com/kubernetes/kubernetes/pull/110319

Comment (ruquanzhao) [reply pkg/util/proto/openapi_suite_test.go:49]: Thanks! I will remove it

**Comment by ruquanzhao** [comment]
> You can rebase this on top of master and you should be good to go.

@apelisse  Thanks! I have rebased this. Need your help to trigger CI again.

**Comment by chendave** [comment]
@apelisse @liggitt  is this good to go?

**Comment by apelisse** [comment]
Thanks!
/lgtm
/approve

**Comment by pohly** [comment]
I think the original reason for putting this on hold has been resolved: ginkgo/v1 no longer appears in go.sum.

/hold cancel

**Pull request merged**